### PR TITLE
feat: add advenced resolver for new semantic modle

### PR DESCRIPTION
**Pull request by NeverRaR**
#### 1. Does this PR affect any open issues?(Y/N) and add issue references (e.g. "fix #123", "re #123".):

- [ ] N
- [x] Y 

re #780 

#### 2. What is the scope of this PR (e.g. component or file name):

kclvm/ast/src/ast.rs
kclvm/ast/src/pos.rs
kclvm/sema/src/advanced_resolver
kclvm/sema/src/core

#### 3. Provide a description of the PR(e.g. more details, effects, motivations or doc link):

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [x] Other

feat: add advenced resolver for new semantic modle

#### 4. Are there any breaking changes?(Y/N) and describe the breaking changes(e.g. more details, motivations or doc link):

- [x] N
- [ ] Y 

#### 5. Are there test cases for these changes?(Y/N) select and add more details, references or doc links:

- [x] Unit test
- [ ] Integration test
- [ ] Benchmark (add benchmark stats below)
- [ ] Manual test (add detailed scripts or steps below)
- [ ] Other

### CODE REVIEW

File: kclvm/sema/src/advanced_resolver/mod.rs [hunk 1]
```diff
@@ -0,0 +1,1038 @@
...
+pub struct Context<'ctx> {
+    pub program: &'ctx Program,
+    node_ty: IndexMap<AstIndex, Rc<Type>>,
+    scopes: Vec<ScopeRef>,
+    current_pkgpath: Option<String>,
+    current_filename: Option<String>,
+    current_schema_symbol: Option<SymbolRef>,
+    start_pos: Position,
+    end_pos: Position,
+    maybe_def: bool,
```
Comment (Peefy) [inline kclvm/sema/src/advanced_resolver/mod.rs:79]: What does the attribute `maybe_def` means (the left and right values)? Would U pls add more documents?

File: kclvm/sema/src/advanced_resolver/mod.rs [hunk 2]
```diff
@@ -0,0 +1,1038 @@
...
+/// Resolver is responsible for program semantic checking, mainly
+/// including type checking and contract model checking.
+pub struct AdvancedResolver<'ctx> {
+    pub(crate) ctx: Context<'ctx>,
+    pub(crate) gs: GlobalState,
+}
+
+pub struct Context<'ctx> {
+    pub program: &'ctx Program,
+    node_ty: IndexMap<AstIndex, Rc<Type>>,
```
Comment (Peefy) [inline kclvm/sema/src/advanced_resolver/mod.rs:72]: s/node_ty/node_ty_mapping

Comment (Peefy) [inline kclvm/sema/src/advanced_resolver/mod.rs:72]: s/Rc<Type>/TypeRef

File: kclvm/sema/src/advanced_resolver/mod.rs [hunk 3]
```diff
@@ -0,0 +1,1038 @@
...
+    },
+    ty::Type,
+};
+use kclvm_ast::ast;
+use kclvm_ast::ast::AstIndex;
+use kclvm_ast::ast::Program;
+use kclvm_ast::walker::MutSelfTypedResultWalker;
+mod node;
+/// Resolver is responsible for program semantic checking, mainly
+/// including type checking and contract model checking.
```
Comment (Peefy) [inline kclvm/sema/src/advanced_resolver/mod.rs:64]: What does the contract model checking means here?

File: kclvm/sema/src/advanced_resolver/mod.rs [hunk 4]
```diff
@@ -0,0 +1,1038 @@
...
+    end_pos: Position,
+    maybe_def: bool,
+}
+
+impl<'ctx> AdvancedResolver<'ctx> {
+    pub fn resolve_program(
+        program: &'ctx Program,
+        gs: GlobalState,
+        node_ty: IndexMap<AstIndex, Rc<Type>>,
+    ) -> GlobalState {
```
Comment (Peefy) [inline kclvm/sema/src/advanced_resolver/mod.rs:87]: Just a small question 🤔 :? Do we need to store error information in this location?

File: kclvm/sema/src/advanced_resolver/mod.rs [hunk 5]
```diff
@@ -0,0 +1,1038 @@
...
+            },
+        };
+
+        for (name, modules) in advanced_resolver.ctx.program.pkgs.iter() {
+            advanced_resolver.ctx.current_pkgpath = Some(name.clone());
+            let pkg_info = advanced_resolver
+                .gs
+                .get_packages()
+                .get_package_info(name)
+                .unwrap();
```
Comment (Peefy) [inline kclvm/sema/src/advanced_resolver/mod.rs:109]: Do not panic in the resolver and the language server.

File: kclvm/sema/src/advanced_resolver/mod.rs [hunk 6]
```diff
@@ -0,0 +1,1038 @@
...
+        advanced_resolver.gs.build_sema_db();
+        advanced_resolver.gs
+    }
+
+    fn enter_root_scope(&mut self, pkgpath: String, filename: String) {
+        let package_ref = self
+            .gs
+            .get_symbols_mut()
+            .get_symbol_by_fully_qualified_name(&pkgpath)
+            .unwrap_or_else(|| panic!("unknown package symbol {pkgpath}"));
```
Comment (Peefy) [inline kclvm/sema/src/advanced_resolver/mod.rs:131]: Do not panic in the resolver and the language server.

File: kclvm/sema/src/advanced_resolver/mod.rs (hunk 4 above)
Comment (NeverRaR) [reply kclvm/sema/src/advanced_resolver/mod.rs:87]: The advanced resolver does not provide  more error information for users. All error checks should be completed in the previous stage of the resolver, and some error information may be added for developer debugging only  in the future.

**Review by Peefy — APPROVED** [review]
LGTM

**Pull request merged**